pkg/libdevfile/libdevfile.go

```diff
@@ -69,6 +87,214 @@ func getDefaultCommand(devfileObj parser.DevfileObj, kind v1alpha2.CommandGroupK
+			commandErrors += debugCmdErr.Error()
+		}
+		return v1alpha2.Command{}, fmt.Errorf(commandErrors)
+	}
+
+	return debugCommand, nil
+}
+
+// ValidateAndGetTestCommands validates the test command
+func ValidateAndGetTestCommands(data data.DevfileData, devfileTestCmd string) (testCommand v1alpha2.Command, err error) {
+	var emptyCommand v1alpha2.Command
+	isTestCommandValid := false
+	testCommand, testCmdErr := GetTestCommand(data, devfileTestCmd)
+	if testCmdErr == nil && !reflect.DeepEqual(emptyCommand, testCommand) {
+		isTestCommandValid = true
+		klog.V(2).Infof("Test command: %v", testCommand.Id)
+	}
+
+	if !isTestCommandValid && testCmdErr != nil {
+		return v1alpha2.Command{}, testCmdErr
+	}
+
+	return testCommand, nil
+}
+
+// GetBuildCommand iterates through the components in the devfile and returns the build command
+func GetBuildCommand(data data.DevfileData, devfileBuildCmd string) (buildCommand v1alpha2.Command, err error) {
+	return getCommand(data, devfileBuildCmd, v1alpha2.BuildCommandGroupKind)
+}
+
+// GetDebugCommand iterates through the components in the devfile and returns the debug command
+func GetDebugCommand(data data.DevfileData, devfileDebugCmd string) (debugCommand v1alpha2.Command, err error) {
+	return getCommand(data, devfileDebugCmd, v1alpha2.DebugCommandGroupKind)
+}
+
+// GetRunCommand iterates through the components in the devfile and returns the run command
+func GetRunCommand(data data.DevfileData, devfileRunCmd string) (runCommand v1alpha2.Command, err error) {
+	return getCommand(data, devfileRunCmd, v1alpha2.RunCommandGroupKind)
+}
+
+// GetTestCommand iterates through the components in the devfile and returns the test command
+func GetTestCommand(data data.DevfileData, devfileTestCmd string) (runCommand v1alpha2.Command, err error) {
+	return getCommand(data, devfileTestCmd, v1alpha2.TestCommandGroupKind)
+}
+
+// ShouldExecCommandRunOnContainer returns whether the given exec command should run on the specified containerName.
+func ShouldExecCommandRunOnContainer(exec *v1alpha2.ExecCommand, containerName string) bool {
+	return exec != nil && exec.Component == containerName
+}
+
+// getCommand iterates through the devfile commands and returns the devfile command associated with the group
+// commands mentioned via the flags are passed via commandName, empty otherwise
+func getCommand(data data.DevfileData, commandName string, groupType v1alpha2.CommandGroupKind) (supportedCommand v1alpha2.Command, err error) {
+
+	var command v1alpha2.Command
+
+	if commandName == "" {
+		command, err = getCommandAssociatedToGroup(data, groupType)
+	} else if commandName != "" {
+		command, err = getCommandFromFlag(data, groupType, commandName)
+	}
+
+	return command, err
+}
+
+// getCommandAssociatedToGroup iterates through the devfile commands and returns the command associated with the group
+func getCommandAssociatedToGroup(data data.DevfileData, groupType v1alpha2.CommandGroupKind) (v1alpha2.Command, error) {
```

Comment:
As I understand it, `getCommandAssociatedToGroup` either returns the default command for this group or the first non-default one it found.
On the other hand, `GetDefaultCommand` returns an error if there is no default command for the specified group. It also checks that there is not more than one default command found for this group while iterating.
So they seem to behave differently. But I agree that they look similar.

Reply:
I think we should try to haromize the behaviour between commands. Deploy is using `getDefaultCommand`, when Build/Test/Run is using `getCommandAssociatedToGroup`

Reply:
Yes, makes sense. I had the same feeling when moving all those functions to `libdevfile`, but wanted to refactor them when working on supporting non-default build and run commands (either #5775 or #5776), since these functions will need to be updated for these issues. WDYT?

Reply:
Yes, makes sense to work on it during #5775 or #5776

Reply:
Okay, I added a comment to these issues to keep track of this point. Thanks.